### PR TITLE
Problem: devnet configs in integration tests are repetitive

### PR DESCRIPTION
**Pull request by mmsqe**
Closes: https://github.com/crypto-org-chain/cronos/issues/554
ref: https://github.com/crypto-com/pystarport/pull/57, bump up dependency

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: integration_tests/configs/utils.jsonnet [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  validator(mnemonic):: {
+    coins: '1000000000000000000stake,10000000000000000000000basetcro',
+    staked: '1000000000000000000stake',
+    mnemonic: mnemonic,
+  },
+  validator_with_timeout(mnemonic):: self.validator(mnemonic) {
+    config: {
```
Comment (yihuang) [inline integration_tests/configs/utils.jsonnet:8]: it's supported to set common configs in upper level now: https://github.com/crypto-com/pystarport/pull/56

Comment (mmsqe) [reply integration_tests/configs/utils.jsonnet:8]: do you mean merge with validator instead of replace whole validators in jsonnet?

File: integration_tests/configs/utils.jsonnet [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  validator(mnemonic):: {
+    coins: '1000000000000000000stake,10000000000000000000000basetcro',
+    staked: '1000000000000000000stake',
+    mnemonic: mnemonic,
+  },
+  validators(mnemonics):: [self.validator(mnemonic) for mnemonic in mnemonics],
+  accounts(objs):: [obj for obj in objs],
```
Comment (yihuang) [inline integration_tests/configs/utils.jsonnet:8]: https://github.com/crypto-com/pystarport/pull/57#pullrequestreview-1024207540
similar opinions on this.

Comment (mmsqe) [reply integration_tests/configs/utils.jsonnet]: removed

File: integration_tests/pyproject.toml [hunk 3]
```diff
@@ -20,10 +20,11 @@ python-dateutil = "^2.8.1"
 web3 = "^5.20.1"
 eth-bloom = "^1.0.4"
 python-dotenv = "^0.19.2"
-pystarport = "^0.2.3"
+pystarport = { git = "https://github.com/mmsqe/pystarport.git", branch = "jsonnet-config-simplify" }
```
Comment (yihuang) [inline integration_tests/pyproject.toml:23]: can use main branch now.

Comment (mmsqe) [reply integration_tests/pyproject.toml:23]: yes, but not sure if ci can cache with git url since current build take ages

Comment (yihuang) [reply integration_tests/pyproject.toml:23]: https://github.com/crypto-org-chain/cronos/pull/558/files#diff-b9a0014cf6c4615d1c2e8d921dc3db59b4a68b462788d667ed622331660d17cdR69
that's becuase you updated `nixpkgs`, you can revert those changes, then it'll be much more cache friendlier.

File: integration_tests/configs/cosmovisor.jsonnet [hunk 4]
```diff
@@ -0,0 +1,31 @@
+local config = import 'default.jsonnet';
+
+config {
+  'cronos_777-1'+: {
+    'app-config'+: {
+      'minimum-gas-prices': '5000000000000basetcro',
+      'json-rpc'+: {
+        api:: super['api'],
```
Comment (yihuang) [inline integration_tests/configs/cosmovisor.jsonnet:8]: no need to override `api` field?

File: integration_tests/configs/default.jsonnet [hunk 5]
```diff
@@ -0,0 +1,86 @@
+{
+  dotenv: '../../scripts/.env',
```
Comment (yihuang) [inline integration_tests/configs/default.jsonnet:2]: Should we replace `dotenv` trick with jsonnet expressions as well?

File: integration_tests/configs/disable_auto_deployment.jsonnet [hunk 6]
```diff
@@ -0,0 +1,30 @@
+local config = import 'default.jsonnet';
+
+config {
+  'cronos_777-1'+: {
+    'start-flags': '--trace --unsafe-experimental',
+    'app-config'+: {
+      'minimum-gas-prices':: super['minimum-gas-prices'],
+      'json-rpc'+: {
+        api:: super['api'],
```
Comment (yihuang) [inline integration_tests/configs/disable_auto_deployment.jsonnet:9]: do we need to override the `api` field?

File: integration_tests/configs/cosmovisor.jsonnet (hunk 4 above)
Comment (mmsqe) [reply integration_tests/configs/cosmovisor.jsonnet:8]: `::` mean remove, no api under [json-rpc](https://github.com/crypto-org-chain/cronos/blob/main/integration_tests/configs/cosmovisor.yaml#L9) before

File: integration_tests/configs/default.jsonnet (hunk 5 above)
Comment (mmsqe) [reply integration_tests/configs/default.jsonnet:2]: not sure if you bother env parser itself? since read env logic is [fixed](https://github.com/crypto-com/pystarport/pull/52/commits/e1f67e2424415c08f637a20203b40dacbebaf8c5)

File: integration_tests/configs/yaml_doc.jsonnet [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// jsonnet yaml_doc.jsonnet -m . -S
```
Comment (yihuang) [inline integration_tests/configs/yaml_doc.jsonnet:1]: Is this file needed?

File: integration_tests/configs/cosmovisor.jsonnet (hunk 4 above)
Comment (yihuang) [reply integration_tests/configs/cosmovisor.jsonnet:8]: I see 👍

File: integration_tests/configs/default.jsonnet (hunk 5 above)
Comment (yihuang) [reply integration_tests/configs/default.jsonnet:2]: I just think it's simple enough with jsonnet, and potentially deprecate dotenv feature in the future, so keep only one way to do the same thing.

File: integration_tests/configs/yaml_doc.jsonnet (hunk 7 above)
Comment (mmsqe) [reply integration_tests/configs/yaml_doc.jsonnet:1]: no harms to keep, can use to generate equivalent yaml

File: integration_tests/configs/default.jsonnet (hunk 5 above)
Comment (yihuang) [reply integration_tests/configs/default.jsonnet:2]: > not sure if you bother env parser itself? since read env logic is [fixed](https://github.com/crypto-com/pystarport/pull/52/commits/e1f67e2424415c08f637a20203b40dacbebaf8c5)

I see, do you mean dotenv provides env reading ability, which doesn't exist in jsonnet approach?

Comment (mmsqe) [reply integration_tests/configs/default.jsonnet:2]: yes

**Review by yihuang — APPROVED** [review]
Good stuff, LGTM.

The only thing is `integration_tests/configs/yaml_doc.jsonnet` seems not useful, if what's needed is a cli to generate jsonnet to yaml for debugging purpose, can we add one and put in `scripts`?

File: integration_tests/configs/yaml_doc.jsonnet (hunk 7 above)
Comment (yihuang) [reply integration_tests/configs/yaml_doc.jsonnet:1]: can we add a cli script to convert jsonnet to yaml, so we don't need to keep this?

**Comment by mmsqe** [comment]
> can we add one and put in `scripts`?

no need for test, I add as a doc, and we need `std.manifestYamlDoc(import './***.jsonnet', true, false)` file anyway, I just think easier do within one file than [generate-test-yamls](https://github.com/crypto-com/pystarport/pull/57/commits/463a136cf3d92948e5cb5c39f59a693e260c6775#diff-ff4382ca3ea45f28e2928f9841026be7d4aa891a5dc78ca0ac32080beb9c237b)

**Comment by yihuang** [comment]
> > can we add one and put in `scripts`?
> 
> no need for test, I add as a doc, and we need `std.manifestYamlDoc(import './***.jsonnet', true, false)` file anyway, I just think easier do within one file than [generate-test-yamls](https://github.com/crypto-com/pystarport/pull/57/commits/463a136cf3d92948e5cb5c39f59a693e260c6775#diff-ff4382ca3ea45f28e2928f9841026be7d4aa891a5dc78ca0ac32080beb9c237b)

Can you make is more generic though, don't need to add an item when adding/removing config files.

**Comment by yihuang** [comment]
> > can we add one and put in `scripts`?
> 
> no need for test, I add as a doc, and we need `std.manifestYamlDoc(import './***.jsonnet', true, false)` file anyway, I just think easier do within one file than [generate-test-yamls](https://github.com/crypto-com/pystarport/pull/57/commits/463a136cf3d92948e5cb5c39f59a693e260c6775#diff-ff4382ca3ea45f28e2928f9841026be7d4aa891a5dc78ca0ac32080beb9c237b)

it's only for debugging, right?

**Comment by mmsqe** [comment]
> it's only for debugging, right?

yes, and a bit hard for jsonnet to include dynamic file

File: integration_tests/configs/generate-test-yamls [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+set +e
+
+name=$1
+sed -e "s/\${file}/${name}/" -e "s/\${file}/${name}/" yaml_doc.jsonnet.template > yaml_doc.jsonnet && jsonnet yaml_doc.jsonnet -m . -S
```
Comment (yihuang) [inline integration_tests/configs/generate-test-yamls:6]:
```suggestion
echo "{'${1%.*}.yaml': std.manifestYamlDoc(import '$1', true, false)}" | jsonnet - -m . -S
```

Comment (yihuang) [reply integration_tests/configs/generate-test-yamls:6]: It seems equivalent to one liner:
```
jsonnet long_timeout_commit.jsonnet | yq -P
```

**Review by yihuang — APPROVED** [review]
LGTM

File: integration_tests/test_mempool.py [hunk 9]
```diff
@@ -61,24 +62,24 @@ def test_mempool(cronos_mempool):
     print(f"block number start: {block_num_0}")
     nonce_begin = w3.eth.get_transaction_count(address_from)
 
-    sended_hash_set = set()
+    txs_to_send = []
     for i in range(5):
         nonce = nonce_begin + i
-        tx = {
+        txs_to_send.append({
```
Comment (yihuang) [inline integration_tests/test_mempool.py:68]: but to do it in batch, it's not really a mempool test case anymore, if it was not stable enough, maybe try add a little bit sleep between the loops?

Comment (mmsqe) [reply integration_tests/test_mempool.py]: strange that after 10 blocks still can't find `0x1127bfa542258dea57bb7a6b8cda1041a547097cb3e601d8e481c54a6f815907`, not sure it's confirmed or fail to included
```block number start: 5
all send tx hash: f{HexBytes('0xce12a45ee6e6a3112bf469dc90361106737b65f61af9168d00ae2b3856d74b01'), HexBytes('0x5ed67709eafb00d074d64b015dc27836e593271d0789d22e4b356a69498b3d77'), HexBytes('0x1127bfa542258dea57bb7a6b8cda1041a547097cb3e601d8e481c54a6f815907'), HexBytes('0x1abec2504f8dbdd4d4c9359f9e9810ee44c4c9ab4848fb1d6ba2ed17dcc6feb7'), HexBytes('0xbf28ee609a7758c0592fcf6c86f0d57da55f8b3bb8c383fba0defc99ea44cc0b')}
all pending tx hash after 1 block: [HexBytes('0xbf28ee609a7758c0592fcf6c86f0d57da55f8b3bb8c383fba0defc99ea44cc0b'), HexBytes('0xce12a45ee6e6a3112bf469dc90361106737b65f61af9168d00ae2b3856d74b01'), HexBytes('0x1abec2504f8dbdd4d4c9359f9e9810ee44c4c9ab4848fb1d6ba2ed17dcc6feb7'), HexBytes('0x5ed67709eafb00d074d64b015dc27836e593271d0789d22e4b356a69498b3d77')]
all pending tx hash after 2 block: []
all pending tx hash after 3 block: []
all pending tx hash after 4 block: []
all pending tx hash after 5 block: []
all pending tx hash after 6 block: []
all pending tx hash after 7 block: []
all pending tx hash after 8 block: []
all pending tx hash after 9 block: []
all pending tx hash after 10 block: []
```
